File: tools/kvmexit.py
#!/usr/bin/env python
#
# kvmexit.py
#
# Display the exit_reason and its statistics of each vm exit
# for all vcpus of all virtual machines. For example:
# $./kvmexit.py
#  PID      TID      KVM_EXIT_REASON                     COUNT
#  1273551  1273568  EXIT_REASON_MSR_WRITE               6
#  1274253  1274261  EXIT_REASON_EXTERNAL_INTERRUPT      1
#  1274253  1274261  EXIT_REASON_HLT                     12
#  ...
#
# Besides, we also allow users to specify one pid, tid(s), or one
# pid and its vcpu. See kvmexit_example.txt for more examples.
#
# @PID: each vitual machine's pid in the user space.
# @TID: the user space's thread of each vcpu of that virtual machine.
# @KVM_EXIT_REASON: the reason why the vm exits.
# @COUNT: the counts of the @KVM_EXIT_REASONS.
#
# REQUIRES: Linux 4.7+ (BPF_PROG_TYPE_TRACEPOINT support)
#
# Copyright (c) 2021 ByteDance Inc. All rights reserved.
#
# 31-Aug-2021  Fei Li <lifei.shirley@bytedance.com>  Initial implementation.
# 28-Jul-2025  Matt Pelland <mpelland@akamai.com>    Implement support for AMD.
# 28-Jul-2025  Matt Pelland <mpelland@akamai.com>    Parallelize postprocessing.
# 28-Jul-2025  Matt Pelland <mpelland@akamai.com>    Silence compiler warnings.

from __future__ import print_function
from time import sleep
from bcc import BPF
import argparse
import multiprocessing
import os
import subprocess

#
# Process Arguments
#
def valid_args_list(args):
    args_list = args.split(",")
    for arg in args_list:
        try:
            int(arg)
        except:
            raise argparse.ArgumentTypeError("must be valid integer")
    return args_list

# arguments
examples = """examples:
    ./kvmexit                              # Display kvm_exit_reason and its statistics in real-time until Ctrl-C
    ./kvmexit 5                            # Display in real-time after sleeping 5s
    ./kvmexit -p 3195281                   # Collpase all tids for pid 3195281 with exit reasons sorted in descending order
    ./kvmexit -p 3195281 20                # Collpase all tids for pid 3195281 with exit reasons sorted in descending order, and display after sleeping 20s
    ./kvmexit -p 3195281 -v 0              # Display only vcpu0 for pid 3195281, descending sort by default
    ./kvmexit -p 3195281 -a                # Display all tids for pid 3195281
    ./kvmexit -p 3195281 -a -m 2           # Display all tids for pid 3195281, limit post processing to two threads
    ./kvmexit -t 395490                    # Display only for tid 395490 with exit reasons sorted in descending order
    ./kvmexit -t 395490 20                 # Display only for tid 395490 with exit reasons sorted in descending order after sleeping 20s
    ./kvmexit -T '395490,395491'           # Display for a union like {395490, 395491}
"""
parser = argparse.ArgumentParser(
    description="Display kvm_exit_reason and its statistics at a timed interval",
    formatter_class=argparse.RawDescriptionHelpFormatter,
    epilog=examples)
parser.add_argument("duration", nargs="?", default=99999999, type=int, help="show delta for next several seconds")
parser.add_argument("-p", "--pid", type=int, help="trace this PID only")
exgroup = parser.add_mutually_exclusive_group()
exgroup.add_argument("-t", "--tid", type=int, help="trace this TID only")
exgroup.add_argument("-T", "--tids", type=valid_args_list, help="trace a comma separated series of tids with no space in between")
exgroup.add_argument("-v", "--vcpu", type=int, help="trace this vcpu only")
exgroup.add_argument("-a", "--alltids", action="store_true", help="trace all tids for this pid")
parser.add_argument("-m", "--max-parallelism", type=int, help="limit post processing parallelism to the given thread count", default=64)
parser.add_argument("-d", "--debug", action="store_true", help="enable debug facilities")
args = parser.parse_args()
duration = int(args.duration)

#
# Setup BPF
#

# load BPF program
bpf_text = """
#include <linux/delay.h>

#define TGID_NUM 1024

struct exit_count {
    u64 exit_ct[REASON_NUM];
};
BPF_PERCPU_ARRAY(init_value, struct exit_count, 1);
BPF_TABLE("percpu_hash", u64, struct exit_count, pcpu_kvm_stat, TGID_NUM);

struct cache_info {
    u64 cache_pid_tgid;
    struct exit_count cache_exit_ct;
};
BPF_PERCPU_ARRAY(pcpu_cache, struct cache_info, 1);

TRACEPOINT_PROBE(kvm, kvm_exit) {
    int cache_miss = 0;
    int zero = 0;
    u32 er = args->exit_reason;
    if (er >= REASON_NUM) {
        return 0;
    }

    u64 cur_pid_tgid = bpf_get_current_pid_tgid();
    u32 tgid = cur_pid_tgid >> 32;
    u32 pid = cur_pid_tgid;

    if (THREAD_FILTER)
        return 0;

    struct exit_count *tmp_info = NULL, *initial = NULL;
    struct cache_info *cache_p;
    cache_p = pcpu_cache.lookup(&zero);
    if (cache_p == NULL) {
        return 0;
    }

    if (cache_p->cache_pid_tgid == cur_pid_tgid) {
        //a. If the cur_pid_tgid hit this physical cpu consecutively, save it to pcpu_cache
        tmp_info = &cache_p->cache_exit_ct;
    } else {
        //b. If another pid_tgid matches this pcpu for the last hit, OR it is the first time to hit this physical cpu.
        cache_miss = 1;

        // b.a Try to load the last cache struct if exists.
        tmp_info = pcpu_kvm_stat.lookup(&cur_pid_tgid);

        // b.b If it is the first time for the cur_pid_tgid to hit this pcpu, employ a
        // per_cpu array to initialize pcpu_kvm_stat's exit_count with each exit reason's count is zero
        if (tmp_info == NULL) {
            initial = init_value.lookup(&zero);
            if (initial == NULL) {
                return 0;
            }

            pcpu_kvm_stat.update(&cur_pid_tgid, initial);
            tmp_info = pcpu_kvm_stat.lookup(&cur_pid_tgid);
            // To pass the verifier
            if (tmp_info == NULL) {
                return 0;
            }
        }
    }

    if (er < REASON_NUM) {
        tmp_info->exit_ct[er]++;
        if (cache_miss == 1) {
            if (cache_p->cache_pid_tgid != 0) {
                // b.*.a Let's save the last hit cache_info into kvm_stat.
                pcpu_kvm_stat.update(&cache_p->cache_pid_tgid, &cache_p->cache_exit_ct);
            }
            // b.* As the cur_pid_tgid meets current pcpu_cache_array for the first time, save it.
            cache_p->cache_pid_tgid = cur_pid_tgid;
            bpf_probe_read(&cache_p->cache_exit_ct, sizeof(*tmp_info), tmp_info);
        }
        return 0;
    }

    return 0;
}
"""

# Defines Intel (VMX) VM exit reason codes. Keep this in sync with
# arch/x86/include/uapi/asm/vmx.h.
VMX_EXIT_REASONS = {
    0: "EXCEPTION_NMI",
    1: "EXTERNAL_INTERRUPT",
    2: "TRIPLE_FAULT",
    3: "INIT_SIGNAL",
    4: "SIPI_SIGNAL",
    6: "OTHER_SMI",
    7: "INTERRUPT_WINDOW",
    8: "NMI_WINDOW",
    9: "TASK_SWITCH",
    10: "CPUID",
    12: "HLT",
    13: "INVD",
    14: "INVLPG",
    15: "RDPMC",
    16: "RDTSC",
    18: "VMCALL",
    19: "VMCLEAR",
    20: "VMLAUNCH",
    21: "VMPTRLD",
    22: "VMPTRST",
    23: "VMREAD",
    24: "VMRESUME",
    25: "VMWRITE",
    26: "VMOFF",
    27: "VMON",
    28: "CR_ACCESS",
    29: "DR_ACCESS",
    30: "IO_INSTRUCTION",
    31: "MSR_READ",
    32: "MSR_WRITE",
    33: "INVALID_STATE",
    34: "MSR_LOAD_FAIL",
    36: "MWAIT_INSTRUCTION",
    37: "MONITOR_TRAP_FLAG",
    39: "MONITOR_INSTRUCTION",
    40: "PAUSE_INSTRUCTION",
    41: "MCE_DURING_VMENTRY",
    43: "TPR_BELOW_THRESHOLD",
    44: "APIC_ACCESS",
    45: "EOI_INDUCED",
    46: "GDTR_IDTR",
    47: "LDTR_TR",
    48: "EPT_VIOLATION",
    49: "EPT_MISCONFIG",
    50: "INVEPT",
    51: "RDTSCP",
    52: "PREEMPTION_TIMER",
    53: "INVVPID",
    54: "WBINVD",
    55: "XSETBV",
    56: "APIC_WRITE",
    57: "RDRAND",
    58: "INVPCID",
    59: "VMFUNC",
    60: "ENCLS",
    61: "RDSEED",
    62: "PML_FULL",
    63: "XSAVES",
    64: "XRSTORS",
    67: "UMWAIT",
    68: "TPAUSE",
    74: "BUS_LOCK",
    75: "NOTIFY",
    76: "TDCALL"
}

# Defines AMD (SVM) VM exit reason codes. Keep this in sync with
# arch/x86/include/uapi/asm/svm.h.
SVM_EXIT_REASONS = {
    0x0: "READ_CR0",
    0x2: "READ_CR2",
    0x3: "READ_CR3",
    0x4: "READ_CR4",
    0x8: "READ_CR8",
    0x10: "WRITE_CR0",
    0x12: "WRITE_CR2",
    0x13: "WRITE_CR3",
    0x14: "WRITE_CR4",
    0x18: "WRITE_CR8",
    0x20: "READ_DR0",
    0x21: "READ_DR1",
    0x22: "READ_DR2",
    0x23: "READ_DR3",
    0x24: "READ_DR4",
    0x25: "READ_DR5",
    0x26: "READ_DR6",
    0x27: "READ_DR7",
    0x30: "WRITE_DR0",
    0x31: "WRITE_DR1",
    0x32: "WRITE_DR2",
    0x33: "WRITE_DR3",
    0x34: "WRITE_DR4",
    0x35: "WRITE_DR5",
    0x36: "WRITE_DR6",
    0x37: "WRITE_DR7",
    0x40: "EXCP_BASE",
    0x5f: "LAST_EXCP",
    0x60: "INTR",
    0x61: "NMI",
    0x62: "SMI",
    0x63: "INIT",
    0x64: "VINTR",
    0x65: "CR0_SEL_WRITE",
    0x66: "IDTR_READ",
    0x67: "GDTR_READ",
    0x68: "LDTR_READ",
    0x69: "TR_READ",
    0x6a: "IDTR_WRITE",
    0x6b: "GDTR_WRITE",
    0x6c: "LDTR_WRITE",
    0x6d: "TR_WRITE",
    0x6e: "RDTSC",
    0x6f: "RDPMC",
    0x70: "PUSHF",
    0x71: "POPF",
    0x72: "CPUID",
    0x73: "RSM",
    0x74: "IRET",
    0x75: "SWINT",
    0x76: "INVD",
    0x77: "PAUSE",
    0x78: "HLT",
    0x79: "INVLPG",
    0x7a: "INVLPGA",
    0x7b: "IOIO",
    0x7c: "MSR",
    0x7d: "TASK_SWITCH",
    0x7e: "FERR_FREEZE",
    0x7f: "SHUTDOWN",
    0x80: "VMRUN",
    0x81: "VMMCALL",
    0x82: "VMLOAD",
    0x83: "VMSAVE",
    0x84: "STGI",
    0x85: "CLGI",
    0x86: "SKINIT",
    0x87: "RDTSCP",
    0x88: "ICEBP",
    0x89: "WBINVD",
    0x8a: "MONITOR",
    0x8b: "MWAIT",
    0x8c: "MWAIT_COND",
    0x8d: "XSETBV",
    0x8e: "RDPRU",
    0x8f: "EFER_WRITE_TRAP",
    0x90: "CR0_WRITE_TRAP",
    0x91: "CR1_WRITE_TRAP",
    0x92: "CR2_WRITE_TRAP",
    0x93: "CR3_WRITE_TRAP",
    0x94: "CR4_WRITE_TRAP",
    0x95: "CR5_WRITE_TRAP",
    0x96: "CR6_WRITE_TRAP",
    0x97: "CR7_WRITE_TRAP",
    0x98: "CR8_WRITE_TRAP",
    0x99: "CR9_WRITE_TRAP",
    0x9a: "CR10_WRITE_TRAP",
    0x9b: "CR11_WRITE_TRAP",
    0x9c: "CR12_WRITE_TRAP",
    0x9d: "CR13_WRITE_TRAP",
    0x9e: "CR14_WRITE_TRAP",
    0x9f: "CR15_WRITE_TRAP",
    0xa2: "INVPCID",
    0xa5: "BUS_LOCK",
    0xa6: "IDLE_HLT",
    0x400: "NPF",
    0x401: "AVIC_INCOMPLETE_IPI",
    0x402: "AVIC_UNACCELERATED_ACCESS",
    0x403: "VMGEXIT"
}

KVM_EXIT_REASONS_BY_VENDOR = {
    "GenuineIntel": VMX_EXIT_REASONS,
    "AuthenticAMD": SVM_EXIT_REASONS
}


# Do some checks
#
try:
    with open("/proc/cpuinfo", "r") as cpuinfo:
        for line in cpuinfo:
            if line.startswith("vendor_id"):
                cpu_vendor = line.split(":")[1].strip()
                break
    if cpu_vendor not in KVM_EXIT_REASONS_BY_VENDOR:
        raise Exception("CPU vendor not supported: %s" % cpu_vendor)

    # Check if kvm module is loaded
    if os.access("/dev/kvm", os.R_OK | os.W_OK):
        pass
    else:
        raise Exception("Please insmod kvm module to use kvmexit tool.")
except Exception as e:
    raise Exception("Failed to do precondition check, due to: %s." % e)

exit_reasons = KVM_EXIT_REASONS_BY_VENDOR[cpu_vendor]

def find_tid(tgt_dir, tgt_vcpu):
    for tid in os.listdir(tgt_dir):
        path = tgt_dir + "/" + tid + "/comm"
        fp = open(path, "r")
        comm = fp.read()
        if (comm.find(tgt_vcpu) != -1):
            return tid
    return -1

# set process/thread filter
thread_context = ""
header_format = ""
need_collapse = not args.alltids
if args.tid is not None:
    thread_context = "TID %s" % args.tid
    thread_filter = 'pid != %s' % args.tid
elif args.tids is not None:
    thread_context = "TIDS %s" % args.tids
    thread_filter = "pid != " + " && pid != ".join(args.tids)
    header_format = "TIDS     "
elif args.pid is not None:
    thread_context = "PID %s" % args.pid
    thread_filter = 'tgid != %s' % args.pid
    if args.vcpu is not None:
        thread_context = "PID %s VCPU %s" % (args.pid, args.vcpu)
        # transfer vcpu to tid
        tgt_dir = '/proc/' + str(args.pid) + '/task'
        tgt_vcpu = "CPU " + str(args.vcpu)
        args.tid = find_tid(tgt_dir, tgt_vcpu)
        if args.tid == -1:
            raise Exception("There's no v%s for PID %d." % (tgt_vcpu, args.pid))
        thread_filter = 'pid != %s' % args.tid
    elif args.alltids:
        thread_context = "PID %s and its all threads" % args.pid
        header_format = "TID      "
else:
    thread_context = "all threads"
    thread_filter = '0'
    header_format = "PID      TID      "
bpf_text = bpf_text.replace('THREAD_FILTER', thread_filter)
bpf_text = bpf_text.replace('REASON_NUM', str(max(exit_reasons.keys()) + 1))
cflags = []

if not args.debug:
    cflags.append("-w")

b = BPF(text=bpf_text, cflags=cflags)


# header
print("Display kvm exit reasons and statistics for %s" % thread_context, end="")
if duration < 99999999:
    print(" after sleeping %d secs." % duration)
else:
    print("... Hit Ctrl-C to end.")

try:
    sleep(duration)
except KeyboardInterrupt:
    print()


# Currently, sort multiple tids in descending order is not supported.
if (args.pid or args.tid):
    ct_reason = []
    if args.pid:
        tgid_exit = {k: 0 for k in exit_reasons}

# output
print("%s%-35s %s" % (header_format, "KVM_EXIT_REASON", "COUNT"))

pcpu_kvm_stat = b["pcpu_kvm_stat"]
pcpu_cache = b["pcpu_cache"]

def extract_pcpu_kvm_exit_reason_count(args):
    pid_tgid, exit_reason, cpu_num = args
    inner_cpu_cache = pcpu_cache[0][cpu_num]
    cachePIDTGID = inner_cpu_cache.cache_pid_tgid
    # Take priority to check if it is in cache

    if cachePIDTGID == pid_tgid.value:
        return inner_cpu_cache.cache_exit_ct.exit_ct[exit_reason]

    # If not in cache, find from kvm_stat
    return pcpu_kvm_stat[pid_tgid][cpu_num].exit_ct[exit_reason]

cpu_count = multiprocessing.cpu_count()
parallelism = min(max(1, int(cpu_count / 2)), args.max_parallelism)
pool = multiprocessing.Pool(parallelism)

for k, v in pcpu_kvm_stat.items():
    tgid = k.value >> 32
    pid = k.value & 0xffffffff
    for i in exit_reasons.keys():
        sum1 = sum(pool.map(
            extract_pcpu_kvm_exit_reason_count,
            [(k, i, c) for c in range(cpu_count)]
        ))

        if sum1 == 0:
            continue

        if (args.pid and args.pid == tgid and need_collapse):
            tgid_exit[i] += sum1
        elif (args.tid and args.tid == pid):
            ct_reason.append((sum1, i))
        elif not need_collapse or args.tids:
            print("%-8u %-35s %-8u" % (pid, exit_reasons[i], sum1))
        else:
            print("%-8u %-8u %-35s %-8u" % (tgid, pid, exit_reasons[i], sum1))

    # Display only for the target tid in descending sort
    if (args.tid and args.tid == pid):
        ct_reason.sort(reverse=True)
        for i in range(0, len(ct_reason)):
            if ct_reason[i][0] == 0:
                continue
            print("%-35s %-8u" % (exit_reasons[ct_reason[i][1]], ct_reason[i][0]))
        break


# Aggregate all tids' counts for this args.pid in descending sort
if args.pid and need_collapse:
    for k, v in tgid_exit.items():
        ct_reason.append((v, k))
    ct_reason.sort(reverse=True)
    for i in range(0, len(ct_reason)):
        if ct_reason[i][0] == 0:
            continue
        print("%-35s %-8u" % (exit_reasons[ct_reason[i][1]], ct_reason[i][0]))

File: libbpf-tools/kvmexit.c
/* SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */

/*
 * kvmexit  Trace kvm exit and show exit reason stat.
 * Copyright (c) 2021 Hengqi Chen
 *
 * Based on kvmexit(8) from BCC by Fei Li.
 * 21-Sep-2021  Hengqi Chen   Created this.
 */
#define _GNU_SOURCE
#include <argp.h>
#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <sys/types.h>
#include <dirent.h>

#include <bpf/libbpf.h>
#include <bpf/bpf.h>
#include "kvmexit.h"
#include "kvmexit.skel.h"
#include "trace_helpers.h"

#define warn(...) fprintf(stderr, __VA_ARGS__)
#define OUTPUT_ROWS_LIMIT	1024
#define MAX_EXIT_REASON		127

static volatile sig_atomic_t exiting = 0;

static pid_t target_pid = 0;
static pid_t target_tid = 0;
static char *target_tids = NULL;
static int target_vcpu = -1;
static bool trace_by_process = true;
static int duration = 0;

static const char *exit_reason[] = {
	[0] = "EXCEPTION_NMI",
	[1] = "EXTERNAL_INTERRUPT",
	[2] = "TRIPLE_FAULT",
	[3] = "INIT_SIGNAL",
	[4] = "SIPI_SIGNAL",
	[7] = "INTERRUPT_WINDOW",
	[8] = "NMI_WINDOW",
	[9] = "TASK_SWITCH",
	[10] = "CPUID",
	[12] = "HLT",
	[13] = "INVD",
	[14] = "INVLPG",
	[15] = "RDPMC",
	[16] = "RDTSC",
	[18] = "VMCALL",
	[19] = "VMCLEAR",
	[20] = "VMLAUNCH",
	[21] = "VMPTRLD",
	[22] = "VMPTRST",
	[23] = "VMREAD",
	[24] = "VMRESUME",
	[25] = "VMWRITE",
	[26] = "VMOFF",
	[27] = "VMON",
	[28] = "CR_ACCESS",
	[29] = "DR_ACCESS",
	[30] = "IO_INSTRUCTION",
	[31] = "MSR_READ",
	[32] = "MSR_WRITE",
	[33] = "INVALID_STATE",
	[34] = "MSR_LOAD_FAIL",
	[36] = "MWAIT_INSTRUCTION",
	[37] = "MONITOR_TRAP_FLAG",
	[39] = "MONITOR_INSTRUCTION",
	[40] = "PAUSE_INSTRUCTION",
	[41] = "MCE_DURING_VMENTRY",
	[43] = "TPR_BELOW_THRESHOLD",
	[44] = "APIC_ACCESS",
	[45] = "EOI_INDUCED",
	[46] = "GDTR_IDTR",
	[47] = "LDTR_TR",
	[48] = "EPT_VIOLATION",
	[49] = "EPT_MISCONFIG",
	[50] = "INVEPT",
	[51] = "RDTSCP",
	[52] = "PREEMPTION_TIMER",
	[53] = "INVVPID",
	[54] = "WBINVD",
	[55] = "XSETBV",
	[56] = "APIC_WRITE",
	[57] = "RDRAND",
	[58] = "INVPCID",
	[59] = "VMFUNC",
	[60] = "ENCLS",
	[61] = "RDSEED",
	[62] = "PML_FULL",
	[63] = "XSAVES",
	[64] = "XRSTORS",
	[67] = "UMWAIT",
	[68] = "TPAUSE",
	[MAX_EXIT_REASON] = "N/A",
};

const char *argp_program_version = "kvmexit 0.1";
const char *argp_program_bug_address =
	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
const char argp_program_doc[] =
"Trace kvm exit and show exit reason stat.\n"
"\n"
"USAGE: kvmexit [-h] [-p PID] [-t TID] [-a] [DURATION]\n"
"\n"
"EXAMPLES:\n"
"    kvmexit                # Trace kvm exit reason\n"
"    kvmexit 5              # Trace for 5s\n"
"    kvmexit -p 1216        # Only trace pid 1216\n"
"    kvmexit -p 1216 -a     # Only trace pid 1216, display by thread\n"
"    kvmexit -p 1216 -c 0   # Only trace vCPU 0 of pid 1216\n"
"    kvmexit -T 1216,1217   # Trace pid 1216 and 1217\n";

static const struct argp_option opts[] = {
	{ "pid", 'p', "PID", 0, "Process ID to trace" },
	{ "tid", 't', "TID", 0, "Thread ID to trace" },
	{ "tids", 'T', "TIDs", 0, "Comma-separated list of thread id to trace" },
	{ "vcpu", 'c', "vCPU", 0, "Trace this vCPU only" },
	{ "alltids", 'a', NULL, 0, "Display by thread id" },
	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
	{},
};

static error_t parse_arg(int key, char *arg, struct argp_state *state)
{
	static int pos_args = 0;
	char *tid;
	long num;

	switch (key) {
	case 'a':
		trace_by_process = false;
		break;
	case 'p':
		errno = 0;
		num = strtol(arg, NULL, 10);
		if (errno || num <= 0) {
			warn("invalid PID: %s\n", arg);
			argp_usage(state);
		}
		target_pid = num;
		break;
	case 't':
		errno = 0;
		num = strtol(arg, NULL, 10);
		if (errno || num <= 0) {
			warn("invalid TID: %s\n", arg);
			argp_usage(state);
		}
		target_tid = num;
		trace_by_process = false;
		break;
	case 'T':
		if (!arg) {
			warn("No thread ids specified\n");
			argp_usage(state);
		}
		tid = strtok(arg, ",");
		while (tid) {
			num = strtol(tid, NULL, 10);
			if (errno || num <= 0) {
				warn("Invalid tids: %s\n", arg);
				argp_usage(state);
			}
			tid = strtok(NULL, ",");
		}
		target_tids = strdup(arg);
		trace_by_process = false;
		break;
	case 'c':
		errno = 0;
		num = strtol(arg, NULL, 10);
		if (errno || num < 0) {
			warn("invalid vCPU: %s\n", arg);
			argp_usage(state);
		}
		target_vcpu = num;
		trace_by_process = false;
		break;
	case 'h':
		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
		break;
	case ARGP_KEY_ARG:
		errno = 0;
		if (pos_args == 0) {
			duration = strtol(arg, NULL, 10);
			if (errno || duration <= 0) {
				warn("invalid duration: %s\n", arg);
				argp_usage(state);
			}
		} else {
			warn("unrecognized positional argument: %s\n", arg);
			argp_usage(state);
		}
		pos_args++;
		break;
	default:
		return ARGP_ERR_UNKNOWN;
	}
	return 0;
}

static int vcpu_to_tid()
{
	char comm_path[PATH_MAX * 2], path[PATH_MAX], comm[80], cpu[80];
	struct dirent *entry;
	DIR *dir;
	FILE *f;
	int err = -1;

	if (target_vcpu == -1)
		return 0;
	snprintf(cpu, sizeof(cpu), "cpu %d\n", target_vcpu);

	if (!target_pid) {
		warn("No pid specified for vCPU %d\n", target_vcpu);
		return -1;
	}

	snprintf(path, sizeof(path), "/proc/%d/task", target_pid);
	dir = opendir(path);
	if (!dir) {
		warn("Failed to open dir %s\n", path);
		return -1;
	}

	while ((entry = readdir(dir))) {
		if (entry->d_type != DT_DIR ||
		   !strcmp(entry->d_name, "..") ||
		   !strcmp(entry->d_name, "."))
			continue;

		snprintf(comm_path, sizeof(comm_path), "%s/%s/comm", path, entry->d_name);
		f = fopen(comm_path, "r");
		if (!f) {
			warn("Failed to open file %s\n", comm_path);
			break;
		}

		if (fgets(comm, sizeof(comm), f) && strcasestr(comm, cpu) != NULL) {
			target_tid = strtol(entry->d_name, NULL, 10);
			err = 0;
			fclose(f);
			break;
		}
		fclose(f);
	}
	closedir(dir);
	return err;
}

static void sig_int(int signo)
{
	exiting = 1;
}

static int sort_by_count(const void *obj1, const void *obj2)
{
	struct exit_stat *v1 = (struct exit_stat *)obj1;
	struct exit_stat *v2 = (struct exit_stat *)obj2;

	return v2->count - v1->count;
}

static void print_stat(struct kvmexit_bpf *obj)
{
	static struct exit_stat values[OUTPUT_ROWS_LIMIT];
	struct exit_key key, *prev_key = NULL;
	int fd = bpf_map__fd(obj->maps.entries);
	int i, err, rows = 0;

	if (trace_by_process)
		printf("%-7s %-20s %-5s\n", "PID", "EXIT_REASON", "COUNT");
	else
		printf("%-7s %-7s %-20s %-5s\n", "PID", "TID", "EXIT_REASON", "COUNT");

	while (1) {
		err = bpf_map_get_next_key(fd, prev_key, &key);
		if (err) {
			if (errno == ENOENT) {
				break;
			}
			warn("bpf_map_get_next_key failed: %s\n", strerror(errno));
			return;
		}
		err = bpf_map_lookup_elem(fd, &key, &values[rows++]);
		if (err) {
			warn("bpf_map_lookup_elem failed: %s\n", strerror(errno));
			return;
		}
		prev_key = &key;
	}

	qsort(values, rows, sizeof(struct exit_stat), sort_by_count);
	for (i = 0; i < rows; i++) {
		struct exit_stat *v = &values[i];
		if (trace_by_process)
			printf("%-7d %-20s %-5d\n",
			       v->pid, exit_reason[v->exit_reason], v->count);
		else
			printf("%-7d %-7d %-20s %-5d\n",
			       v->pid, v->tid, exit_reason[v->exit_reason], v->count);
	}
}

int main(int argc, char **argv)
{
	static const struct argp argp = {
		.options = opts,
		.parser = parse_arg,
		.doc = argp_program_doc,
	};
	struct kvmexit_bpf *obj;
	int err, map_fd;
	pid_t tid;
	char *p;

	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
	if (err)
		return err;

	err = vcpu_to_tid();
	if (err)
		return err;

	err = bump_memlock_rlimit();
	if (err) {
		warn("failed to increase rlimit: %d\n", err);
		return 1;
	}

	obj = kvmexit_bpf__open();
	if (!obj) {
		warn("failed to open BPF object\n");
		return 1;
	}

	obj->rodata->target_pid = target_pid;
	obj->rodata->target_tid = target_tid;
	obj->rodata->trace_by_process = trace_by_process;
	obj->rodata->filter_by_tid = target_tids != NULL;

	err = kvmexit_bpf__load(obj);
	if (err) {
		warn("failed to load BPF object: %d\n", err);
		goto cleanup;
	}

	if (target_tids) {
		map_fd = bpf_map__fd(obj->maps.tids);
		p = strtok(target_tids, ",");
		while (p) {
			tid = strtol(p, NULL, 10);
			bpf_map_update_elem(map_fd, &tid, &tid, BPF_ANY);
			p = strtok(NULL, ",");
		}
	}

	err = kvmexit_bpf__attach(obj);
	if (err) {
		warn("failed to attach BPF programs: %d\n", err);
		goto cleanup;
	}

	if (signal(SIGINT, sig_int) == SIG_ERR) {
		warn("can't set signal handler: %s\n", strerror(errno));
		goto cleanup;
	}

	printf("Tracing kvm exit reason");
	if (duration)
		printf(" for %d secs.\n", duration);
	else
		printf("... Hit Ctrl-C to end.\n");

	if (duration) {
		sleep(duration);
	} else {
		pause();
		printf("\n");
	}
	print_stat(obj);

cleanup:
	kvmexit_bpf__destroy(obj);

	return err != 0;
}
